.circleci/meta-linter/main.go

```diff
@@ -18,21 +18,104 @@ import (
 	"fmt"
 	"log"
 	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"bitbucket.org/creachadair/stringset"
 )
 
 func main() {
 	errors := []error{}
 
-	// Run each checker and run up the list of errors.
-	errors = append(errors, checkRulesRegistered()...)
-	errors = append(errors, ruleFilename()...)
-	errors = append(errors, checkRulesDocumented()...)
+	// Keep track of rules processed, and which ones have one or more failures.
+	pass := stringset.New()
+	fail := stringset.New()
+
+	// Begin walking the rules directory looking for rules.
+	err := filepath.Walk("./rules", func(path string, info os.FileInfo, err error) error {
+		// Sanity check: Bubble errors.
+		if err != nil {
+			errors = append(errors, err)
+			return nil
+		}
+
+		// Weed out files that are not rules (tests, etc.).
+		for _, exempt := range []func(path string, info os.FileInfo) bool{
+			func(_ string, i os.FileInfo) bool { return i.IsDir() },
+			func(p string, _ os.FileInfo) bool { return strings.Contains(p, "/internal/") },
+			func(p string, _ os.FileInfo) bool { return p == "rules/rules.go" },
+			func(p string, _ os.FileInfo) bool { return strings.HasSuffix(p, "_test.go") },
+			func(p string, _ os.FileInfo) bool { return aipIndex.MatchString(p) },
+		} {
+			if exempt(path, info) {
+				return nil
+			}
+		}
+
+		// This represents a rule. Get the AIP and rule name from the path.
+		match := ruleFile.FindStringSubmatch(path)
+		if match == nil {
+			errors = append(errors, fmt.Errorf("unexpected path: %s", path))
+			return nil
+		}
+
+		// Get the AIP number and final rule segment.
+		aip, err := strconv.Atoi(match[1])
+		if err != nil {
+			errors = append(errors, err)
+			return nil
+		}
+		name := strings.ReplaceAll(match[2], "_", "-")
+		token := fmt.Sprintf("%04d-%s", aip, name)
+
+		// Run each checker and run up the list of errors.
+		for _, checker := range []func(aip int, name string) []error{
```

Comment:
Can we return an error immediately if found in a checking, to increase code readability? Otherwise, we need to keep tracking the error array mentally.

Reply:
Sure. I do not think it matters much, but I have no reason not to do it.

Reply:
Actually, it turns out I do not want to do this. The current code handles the boundary conditions (e.g. unexpected filenames) better. I _can_ do what you suggest, by using some kind of global tracking failure bit and other scaffolding, but it turns out not to be an improvement. :(

Reply:
Fair enough!